pkg/ipam-controller/allocator/allocator.go

```diff
@@ -0,0 +1,386 @@
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	nodeAllocations := make(map[string]*pool.IPPool, len(a.allocators))
+	for poolName, allocator := range a.allocators {
+		allocation, err := allocator.Allocate(ctx, nodeName)
+		if err != nil {
+			a.deallocate(ctx, nodeName)
+			return nil, err
+		}
+		nodeAllocations[poolName] = nodeAllocationInfoToIPPoolConfig(poolName, allocation)
+	}
+
+	if log.V(1).Enabled() {
+		//nolint:errchkjson
+		dump, _ := json.Marshal(nodeAllocations)
+		log.V(1).Info("allocated ranges", "ranges", dump)
+	}
+	return nodeAllocations, nil
+}
+
+func (a *Allocator) deallocate(ctx context.Context, nodeName string) {
+	for _, allocator := range a.allocators {
+		allocator.Deallocate(ctx, nodeName)
+	}
+}
+
+// Deallocate release all ranges allocated for node
+func (a *Allocator) Deallocate(ctx context.Context, nodeName string) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	a.deallocate(ctx, nodeName)
+}
+
+func (a *Allocator) configure(ctx context.Context, configs []AllocationConfig) {
+	log := logr.FromContextOrDiscard(ctx)
+	for _, cfg := range configs {
+		poolLog := log.WithValues("pool", cfg.PoolName,
+			"gateway", cfg.Gateway.String(), "subnet", cfg.Subnet.String(), "perNodeBlockSize", cfg.PerNodeBlockSize)
+		pAlloc, exist := a.allocators[cfg.PoolName]
+		if exist {
+			poolLog.Info("update IP pool allocator config")
+			pAlloc.Configure(ctx, cfg)
+		} else {
+			poolLog.Info("initialize IP pool allocator")
+			a.allocators[cfg.PoolName] = newPoolAllocator(cfg)
+		}
+	}
+	// remove outdated pools from controller state
+	for poolName := range a.allocators {
+		found := false
+		for _, cfg := range configs {
+			if poolName == cfg.PoolName {
+				found = true
+				break
+			}
+		}
+		if !found {
+			delete(a.allocators, poolName)
+		}
+	}
+}
+
+func nodeAllocationInfoToIPPoolConfig(poolName string, alloc nodeAllocationInfo) *pool.IPPool {
+	return &pool.IPPool{
+		Name:    poolName,
+		Subnet:  alloc.Subnet.String(),
+		StartIP: alloc.StartIP.String(),
+		EndIP:   alloc.EndIP.String(),
+		Gateway: alloc.Gateway.String(),
+	}
+}
+
+func ipPoolConfigToNodeAllocationInfo(node string, alloc *pool.IPPool) (nodeAllocationInfo, error) {
+	if alloc == nil {
+		return nodeAllocationInfo{}, fmt.Errorf("node allocation is nil")
+	}
+	_, subnet, err := net.ParseCIDR(alloc.Subnet)
+	if subnet == nil || err != nil {
+		return nodeAllocationInfo{}, fmt.Errorf("subnet is incorrect network")
+	}
+	gateway := net.ParseIP(alloc.Gateway)
+	if gateway == nil {
+		return nodeAllocationInfo{}, fmt.Errorf("gateway is incorrect ip")
+	}
+	nodeAllocStart := net.ParseIP(alloc.StartIP)
+	if nodeAllocStart == nil {
+		return nodeAllocationInfo{}, fmt.Errorf("startIP is incorrect ip")
+	}
+	nodeAllocEnd := net.ParseIP(alloc.EndIP)
+	if nodeAllocEnd == nil {
+		return nodeAllocationInfo{}, fmt.Errorf("endIP is incorrect ip")
+	}
+
+	if !subnet.Contains(nodeAllocStart) || !subnet.Contains(nodeAllocEnd) {
```

Comment:
you don't validate that the gateway is in the subnet.

Reply:
fixed